### DISCUSSION TITLE
OpenGL control doesn't render in macOS

### DISCUSSION THREAD

**Discussion by danipen** [AvaloniaUI/Avalonia#16547] [Q&A]
I'm trying the latest master ([3deddbe](https://github.com/AvaloniaUI/Avalonia/commit/3deddbe3050f67d2819d1710b2f1062b7b15868e)) and both "OpenGL" and "OpenGL Lease" tabs display empty contents.
![image](https://github.com/user-attachments/assets/e9fd4e19-caa4-4b44-8b74-1d528d48f43c)

I'm using macOS Sonoma 14.5 (23F79), 2,4 GHz 8-Core Intel Core i9, AMD Radeon Pro 5500M 8 GB
Any ideas?

**Comment by BAndysc** [comment]
The latest master uses Metal by default and OpenGlControl doesn't work with Metal yet. You can change it back to OpenGL

```
.With(new AvaloniaNativePlatformOptions() { RenderingMode = new[] { AvaloniaNativeRenderingMode.OpenGl, AvaloniaNativeRenderingMode.Software } })
```

**Comment by danipen** [comment]
@maxkatz6 @kekekeks are there any plans for making metal renderer to support OpenGlControl?

**Reply by BAndysc** [reply]
https://github.com/AvaloniaUI/Avalonia/issues/12728#issuecomment-1700432353
I don't know if there is any update on this.

**Reply by maxkatz6** [reply]
Yes, we want to support opengl with metal. But we are not working on that yet.

**Reply by danipen** [reply]
Do you have any estimated time and task size? I guess no...but I'm asking to figure out when this could be ready (days, weeks, months, years...). Thanks you!

**Reply by maxkatz6** [reply]
In theory, using CVPixelBufferCreate with metal interopability API should be enough. https://developer.apple.com/documentation/metal/metal_sample_code_library/mixing_metal_and_opengl_rendering_in_a_view?language=objc

I have some basic metal experience and can try to take a look, but my OpenGL experience isn't really better to be confident :)

**Reply by maxkatz6** [reply]
Okay, it's more complicated than that. [This PR](https://github.com/AvaloniaUI/Avalonia/pull/14653) needs to be completed first, as we don't have any way to create gl context independently from the main rendering backend yet.
